kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.module.kafka.source;
+
+import kafka.serializer.Decoder;
+import kafka.serializer.DefaultDecoder;
+import kafka.utils.VerifiableProperties;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.stream.annotation.Bindings;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.annotation.Bean;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.dsl.kafka.Kafka;
+import org.springframework.integration.kafka.core.*;
+import org.springframework.integration.kafka.listener.KafkaMessageListenerContainer;
+import org.springframework.integration.kafka.listener.KafkaNativeOffsetManager;
+import org.springframework.integration.kafka.listener.OffsetManager;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Source module that receives data from Kafka.
+ *
+ * @author Soby Chacko
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(KafkaConfigurationProperties.class)
+public class KafkaSource {
+
+	@Autowired
+	@Bindings(KafkaSource.class)
+	private Source source;
+
+	@Autowired
+	private KafkaConfigurationProperties properties;
+
+	@Bean
+	public KafkaMessageListenerContainer container() {
+
+		Assert.isTrue(!ObjectUtils.isEmpty(properties.getTopics()) || !ObjectUtils.isEmpty(properties.getPartitions()),
+				"Either a list of topics OR a topic=<comma separated partitions> must be provided");
```

Comment:
I mean something like this:
```
@AssertTrue(message = "exactly one of  'topics' and 'partitions' are mutually exclusive")
public boolean isTopicsOrPartitions() {
	return !ObjectUtils.isEmpty(getTopics()) || !ObjectUtils.isEmpty(getPartitions();
}
```
and place it into the `KafkaConfigurationProperties` and here in the `@Configuration` just do `if...else`.

Although I don't know: maybe we should use something as a default topic?..